diff --git a/gnome-session/leader-sysvinit.c b/gnome-session/leader-sysvinit.c
index 272c3ad..9ebd85c 100644
--- a/gnome-session/leader-sysvinit.c
+++ b/gnome-session/leader-sysvinit.c
@@ -111,6 +111,63 @@ start_script_exited_cb (GPid     pid,
         }
 }
 
+// This is all a hack! DCONF_PROFILE is not set in leader-main.c for
+// OpenRC systems since it relies on the systemd environment.d
+// file. So we have to read this ourselves.
+static void
+try_get_gsetting_gdm_session_name(char const *user, char **session_name)
+{
+        if (g_getenv("DCONF_PROFILE") && !strcmp("gdm", g_getenv("DCONF_PROFILE")))
+        {
+                // but oh, is it a mystery? it's safe to assure the
+                // below is done, as this is mostly all done in
+                // leader-main.c
+                return;
+        }
+        g_autoptr (GError) error = NULL;
+        // HACK: We need DCONF_PROFILE
+        if (!strcmp(*session_name, "gnome") &&
+            (!strcmp(user, "gdm-greeter") || sscanf(user, "gdm-greeter-%*d") > 0))
+        {
+                g_autofree char *filename = g_strdup_printf("/run/gdm/home/%s/.config/environment.d/gdm.conf", user);
+                g_autoptr (GIOChannel) chan = g_io_channel_new_file(filename, "r", &error);
+                if (error)
+                {
+                        g_warning("Couldn't open %s: %s", filename, error->message);
+                        goto out;
+                }
+
+                char *env;
+                gsize len;
+                while (g_io_channel_read_line(chan, &env, &len, NULL, NULL) == G_IO_STATUS_NORMAL)
+                {
+                        // anyone who tells you to use regex is lying to you
+                        char key[64], val[256];
+                        if (sscanf(env, "%63[^=]=%255s", key, val) != 2)
+                                goto out;
+
+                        // yes, this is the _only_ thing we want :)
+                        if (!strcmp(key, "DCONF_PROFILE"))
+                        {
+                                g_debug("Setting %s to %s, the nasty way...", key, val);
+                                setenv(key, val, true);
+                                goto out;
+                        }
+                }
+        }
+out:
+        g_debug("Dconf profile is: %s", g_getenv("DCONF_PROFILE"));
+
+        if (g_getenv("DCONF_PROFILE") && !strcmp(g_getenv("DCONF_PROFILE"), "gdm"))
+        {
+                g_autoptr (GSettings) settings = g_settings_new("org.gnome.desktop.session");
+                *session_name = g_settings_get_string(settings, "session-name");
+                g_info("Hijacked the session-name: %s", *session_name);
+        }
+
+        // we're fine at this point, it's probably just a regular user
+}
+
 static void
 debug_logger (gchar const *log_domain,
               GLogLevelFlags log_level,
@@ -146,7 +203,7 @@ main (int argc, char **argv)
         g_log_set_default_handler (debug_logger, NULL);
         g_autoptr (GError) error = NULL;
         g_auto (Leader) ctx = { .fifo_fd = -1, .start_script_pid = 0 };
-        const char *session_name = NULL;
+        g_autofree char *session_name = NULL;
         const char *debug_string = NULL;
         g_autofree char *session_rundir = NULL;
         g_autofree char *fifo_path = NULL;
@@ -154,13 +211,15 @@ main (int argc, char **argv)
 
         if (argc < 2)
                 g_error ("No session name was specified");
-        session_name = argv[1];
+        session_name = strdup(argv[1]);
 
         char const *user = g_getenv ("USER");
         if (!user)
-                user = "gdm";
+                user = "gdm-greeter";
         g_info ("User is: %s", user);
 
+        try_get_gsetting_gdm_session_name(user, &session_name);
+
         char const *session_type = g_getenv ("XDG_SESSION_TYPE");
         g_info ("XDG_RUNTIME_DIR: %s", g_getenv ("XDG_RUNTIME_DIR"));
 
